fix: update setup instructions

bfdykstra opened the pull request:
## Description

- Please include a summary of the changes and the related issue.

I added documentation to the readme that adds the steps I took to get the app running locally without docker.

I also edited the `run_macos.sh` script to include the installation of dependencies like homebrew and libmagic.

I added `unstructured` to the dependency array of kotaemon pyproject.toml
- Fixes #138

## Type of change

- [x] New features (non-breaking change).
- [x] Bug fix (non-breaking change).
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected).

## Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have added thorough tests if it is a core feature.
- [x] There is a reference to the original bug report and related work.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] The feature is well documented.

Commit:
activate directory to gitignore

Commit:
add my custom env to gitignore, will have to change that

Commit:
add unstructured to kotaemon pyproject.toml

Commit:
Merge branch 'main' of https://github.com/Cinnamon/kotaemon into fixes/setup-instructions

Commit:
add .env to gitignore

Commit:
remove .env from tracking

Commit:
make changes to the run_macos script, update readme with more detailed instructions

Commit:
remove my personal changes from gitignore

Commit:
remove line from run_macos script

cin-albert commented on README.md:
```diff
@@ -79,6 +79,12 @@ it is, please follow the non-technical [User Guide](https://cinnamon.github.io/k
+
+1. Python 3.10>=
+2. [homebrew](https://brew.sh/)
+3. [Docker](https://www.docker.com/)
```
These should be optional as we provide options to install from source without docker and homebrew.

bfdykstra replied:
if `unstructured` is going to remain a dependency of `kotaemon`, then it's system dependencies should be documented as well - https://docs.unstructured.io/open-source/introduction/quick-start

cin-albert commented on README.md:
```diff
@@ -108,9 +114,15 @@ cd kotaemon
 pip install -e "libs/ktem"
+
+# optional system dependencies if not installed
+brew install libmagic
```
Could you explain why we need libmagic here? What is it used for?

bfdykstra replied:
It is a system dependency of the `unstructured` package that is used to detect file types. You guys use unstructured to do all of your document parsing. The system dependencies are documented here: https://docs.unstructured.io/open-source/installation/full-installation

cin-albert commented on scripts/run_macos.sh:
```diff
@@ -173,12 +208,17 @@ python_version="3.10"
+    activate_conda_env
+else
+    echo "miniconda installation skipped."
+fi
```
We initially created `scripts/run_*` as installers for non-dev end users. They may not be familiar with Python and conda, and to avoid permission issues, we managed conda ourselves without caring if it already existed system-wide. This way, we didn't require user confirmation to install it. Additionally, miniconda is relatively small, so it didn't add much overhead to the application.

bfdykstra replied:
🤔 🤔 🤔 For sure and I understand the want to support non-dev end users! Do you guys want to make installing miniconda a requirement and not optional? 

This is outside the scope of this PR, but I feel that if you want a non-dev end users to use this application possibly including a clickable `.app`, `.exe` package would be great - Maybe [pyinstaller](https://pyinstaller.org/en/v6.10.0/index.html) could be a good option for that. From my experience (and feel free to disagree with me), non technical users don't want to open up the terminal and type anything in :) it can be very scary.

lone17 replied:
Just wanted to chime in as using conda was partly my decision. The motivation for making miniconda a requirement and not optional was to make the app as self-contained as possible. I wanted to avoid issues where the user's miniconda stopped working or got uninstalled making the app suddenly couldn't start anymore. Yes it would take more space but at the time I deem it as it's worth the tradeoff. I took inspiration from `text-generation-webui` (which also maintains its own conda binary) and `pre-commit` (which maintains its own python env to execute hooks).

> This is outside the scope of this PR, but I feel that if you want a non-dev end users to use this application possibly including a clickable .app, .exe package would be great - Maybe [pyinstaller](https://pyinstaller.org/en/v6.10.0/index.html) could be a good option for that. From my experience (and feel free to disagree with me), non technical users don't want to open up the terminal and type anything in :) it can be very scary.

Yes it would be ideal to have installation files such as `.app` and `.exe`. I love to have it available but at the time we simply didn't have the bandwidth to support it. And having all-in-one scripts was the next best option as it's easier for us to implement and maintain. And I completely agree with you that opening a terminal for a non-dev folks would be an alien experience, so we tried our best to mitigate that:
- the window `.bat` file was already double-click only
- for mac, I included a step-by-step to enable double-clicking the script in [the doc](https://cinnamon.github.io/kotaemon/#installation). I'm not familiar with mac os so this was the best way I could find, please suggest better alternatives if you know any.
- for linux, well, I just hoped that if the user is using Linux he/she would be comfortable with running commands in a terminal 😄

And tbh, we made that months ago and didn't expect it to get this many users. If enough non-dev users are interested in this project I hope we will have more helping hands to develop such clickable installation packages. (back then we even struggled to find a mac to test the scripts 😅) Now, whether to have that in the roadmap is entirely up to the current team. If you feel that it should be prioritized, feel free to create a feature request, or even better, contribute a PR, I'm sure it will be highly appreciated 😊

bfdykstra replied:
Sounds good and totally reasonable. I'll change this so it doesn't ask for now and instead just installs it

You can change that if you'd like to have a clickable file for non-dev users on mac :)

cin-albert commented on scripts/run_macos.sh:
```diff
@@ -87,8 +87,39 @@ function deactivate_conda_env() {
+
+    echo "Installing system dependencies...."
+    check_homebrew_installed
+    brew install libmagic
```
Let's verify the need for libmagic and homebrew

bfdykstra replied:
Yup, see comment above re `unstructured` system dependencies. As for homebrew, it's fairly standard practice to use it on a Mac as an installer for system packages but I'll leave that decision up to you all!

cin-albert reviewed (commented):
Hi @bfdykstra,
Thanks so much for your interest in our work.
I've reviewed your changes and have a few thoughts. Feel free to address my comments and discuss these points further

bfdykstra commented:
Thank you for the feedback @cin-albert ! I addressed your comments as best I could. Basically I think that the dependencies of `unstructured` need to be documented in order for any file upload and parsing to work properly.

Additionally, (and this could be a separate issue, does not need to be addressed here) I think it would be advantageous to include a clickable package or release that non technical users could run without opening their terminal.

Commit:
remove option for not installing miniconda for non technical users, mark docker dependency as optional

Commit:
Merge branch 'main' of https://github.com/Cinnamon/kotaemon into fixes/setup-instructions

taprosoft committed:
docs: update demo URL

Commit:
Merge branch 'main' of https://github.com/Cinnamon/kotaemon into fixes/setup-instructions

Niko-La commented:
@bfdykstra the instruction `ollama pull llama3.1:8b` for model 

but  for LLM settings
```
api_key: ollama
base_url: http://localhost:11434/v1/
model: gemma2:2b (for llm) | nomic-embed-text (for embedding)
```

Better to show `ollama pull gemma2:2b`  to keep it consistent ? perfer to give gemma2 as its smalest model to run on most lower end hardware.

bfdykstra commented:
> @bfdykstra the instruction `ollama pull llama3.1:8b` for model
> 
> but for LLM settings
> 
> ```
> api_key: ollama
> base_url: http://localhost:11434/v1/
> model: gemma2:2b (for llm) | nomic-embed-text (for embedding)
> ```
> 
> Better to show `ollama pull gemma2:2b` to keep it consistent ? perfer to give gemma2 as its smalest model to run on most lower end hardware.

@Niko-La throughout most of the docs it says to pull `llama3.1:8b` so there are inconsistencies everywhere. For example [here](https://github.com/Cinnamon/kotaemon/blob/main/docs/local_model.md#ollama-openai-compatible-server-recommended) it says to pull the `llama3.1` and then immediately below it says to use `gemma`.... I want to keep the scope of this PR just to the initial readme and to fix the dependencies so that users can get started with default openai

Commit:
Merge branch 'main' of https://github.com/Cinnamon/kotaemon into fixes/setup-instructions

Commit:
Merge branch 'main' of https://github.com/Cinnamon/kotaemon into fixes/setup-instructions

Commit:
gitignore changes

bfdykstra commented:
@lone17 or @cin-albert Is there anything I should change here?

cin-albert commented:
@bfdykstra, Sorry it took me so long to get back to you.

Agree with you that Unstructured should be treated as a mandatory dependency for Kotaemon and homebrew is used to install system requirements. However, according to Unstructured's document, we might need to install other system dependencies as well, such as `libreoffice` and `tesseract`. Could you help with this?

FYI, Kotaemon uses Unstructured for a wide variety of doc types. We'll need pyproject.toml to install extra required for those documents, for example, `pip install unstructured[pdf]` for PDFs
https://github.com/Cinnamon/kotaemon/blob/73a476979e96b1475d409dfc7542fe0603856e44/libs/kotaemon/kotaemon/indices/ingests/files.py#L38-L54

lone17 commented:
@bfdykstra @cin-albert cc @taprosoft 
Hi, I have some concerns about making `unstructured` a mandatory dependency. As it needs quite a few extra system dependencies, I think we should consider it as an advanced feature only. What I mean by that is:
- not include it in the default dependencies, but instead in the `adv` section https://github.com/Cinnamon/kotaemon/blob/73a476979e96b1475d409dfc7542fe0603856e44/libs/kotaemon/pyproject.toml#L64-L73
- not installing it in the run_*.sh scripts, accepting the fact that we cannot ship it with the non-tech installation route until we have something like #48 
- include it in the doc for developers
- include it in the docker installation
- automatically fall back to a different reader when `unstructured` is not installed, or gracefully show an error message https://github.com/Cinnamon/kotaemon/blob/73a476979e96b1475d409dfc7542fe0603856e44/libs/kotaemon/kotaemon/indices/ingests/files.py#L38-L54

bfdykstra commented:
Is the package as it stands right now runnable and/or working without unstructured installed?

If docker container install gets working from #274 then this PR may be redundant 

My only goal here is to make it easier for people to get up and running :)

lone17 commented:
> If docker container install gets working from #274 then this PR may be redundant
> 
> My only goal here is to make it easier for people to get up and running :)
Oh no it's not redundant since we still need to update the readme :)

I'm unsure about the docker installation, would need @phv2312's confirmation on this one.

> Is the package as it stands right now runnable and/or working without unstructured installed?

It used to be. But I haven't tried it since the big `v0.4.0` update, can you help confirm this point @cin-albert  ?

cin-albert commented:
@lone17 
>  As it needs quite a few extra system dependencies, I think we should consider it as an advanced feature only

I agree that Unstructured is complicated to install in our bash/batch scripts because it requires some system dependencies and will involve system permission issues. However, if we omit unstructured in the installer, it could conflict with our original value of making Kotaemon friendly to non-dev users.

Yes, the best solution is to fall back to another indexing method in case Unstructured is not available, but for now, we have not prepared them so Unstructured is still an essential component.

In the short term, this is the simplest solution, which can satisfy the community, at least for MacOS users with homebrew that does not require sudo. In the long term, we are working on a better package management method (since we have many advanced features, and people tend to add more dependencies to this project), which could address the problem of Unstructured.

Nice to hear your perspectives.

@bfdykstra 
> Is the package as it stands right now runnable and/or working without unstructured installed?

Yes, Kotaemon can work without Unstructured, but you can only ingest `.pdf`, ` .html`, `.mhtml`, and '`.xlsx` documents.

> If docker container install gets working from https://github.com/Cinnamon/kotaemon/pull/274 then this PR may be redundant

I don't think your PR overlaps with the Docker update PR. There are still a large amount of users preferring the bash/batch installers over docker.

cin-albert commented:
@bfdykstra @lone17 
After discussing with the maintenance team, I would like to change my opinion: Unstructured is still treated as optional because of the heavy installation. We will add methods to allow users to install advanced features in the future.

Sorry for the confusion, and thanks @bfdykstra for your contributions. Please update your PR accordingly, or if you think it's unnecessary now, I would like to close the PR

bfdykstra commented:
Okay thank you for the careful consideration!

To summarize: 
1. I will revert the changes the to `run_macos.sh` script
2. The `unstructured` dependency will go under the `adv` array in `pyproject.toml`
3. In the Readme under "For developers" section, I will add in instructions and links to the system dependencies if they would like to process files other than `.pdf`,  `.htm`l, `.mhtml`, and `.xlsx` documents.

Anything I missed or that you would add?

cin-albert commented:
@bfdykstra 
That's all we need for now. Also, please double-check the pre-commit style check failure

lone17 commented:
@bfdykstra 
Thank you for your contribution and sorry for having you edit the run script in vain. 🙏🏼

Commit:
merge .env.example

Commit:
Merge branch 'main' of https://github.com/Cinnamon/kotaemon into fixes/setup-instructions

Commit:
revert changes to run_macos.sh

Commit:
unstructured to advanced dependencies

Commit:
add link to unstructured system dependencies

Commit:
remove api key

bfdykstra commented:
@cin-albert @lone17 All set now - take a look?

bfdykstra committed:
Merge branch 'main' into fixes/setup-instructions

cin-albert commented:
LGTM now, but let me fix the Lint check for commit messages before merge.

Btw, please ensure your secret is rotated.

bfdykstra commented:
Thank you!

Yes I rotated it 🤦‍♂️That’s the danger of having .env tracked. Thank you
for checking

On Tue, Sep 17, 2024 at 12:59 AM Quang (Albert) ***@***.***>
wrote:

> LGTM now, but let me fix the Lint check for commit messages before merge.
>
> Btw, please ensure your secret is rotated.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Cinnamon/kotaemon/pull/144#issuecomment-2354814845>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACP3ZFW2P2V4FCB3VQ2TVUDZW7OPDAVCNFSM6AAAAABNJBDPOOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNJUHAYTIOBUGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cin-albert committed:
fix: skip tests when unstructured pdf not installed

cin-albert committed:
chore: loosen unstructured package version in pyproject.toml

cin-albert reviewed (dismissed):
@bfdykstra 
Allow me to make some changes to your PR to comfort the tests. Feel free to revert or reset them if you feel unsatisfied

cin-albert committed:
chore: correct syntax

lone17 commented:
LGTM. I don't have reviewer privilege though :))

cin-albert committed:
Merge branch 'main' into fixes/setup-instructions

taprosoft commented on libs/kotaemon/pyproject.toml:
```diff
@@ -73,11 +73,14 @@ adv = [
     "llama-index>=0.10.40,<0.11.0",
     "llama-index-vector-stores-milvus",
     "llama-index-vector-stores-qdrant",
+    "python-docx>=1.1.0,<1.2",
```
remove duplicated dependency

taprosoft reviewed (approved):
LGTM

taprosoft commented:
Thanks all @bfdykstra @cin-albert @lone17. Sorry for the long wait.
